engine/apps/shift_swaps/models/shift_swap_request.py

```diff
@@ -0,0 +1,140 @@
+import typing
+
+from django.conf import settings
+from django.core.validators import MinLengthValidator
+from django.db import models
+from django.utils import timezone
+
+from apps.shift_swaps import exceptions
+from common.public_primary_keys import generate_public_primary_key, increase_public_primary_key_length
+
+if typing.TYPE_CHECKING:
+    from apps.user_management.models import User
+
+
+def generate_public_primary_key_for_shift_swap_request() -> str:
+    prefix = "SSR"
+    new_public_primary_key = generate_public_primary_key(prefix)
+
+    failure_counter = 0
+    while ShiftSwapRequest.objects.filter(public_primary_key=new_public_primary_key).exists():
+        new_public_primary_key = increase_public_primary_key_length(
+            failure_counter=failure_counter, prefix=prefix, model_name="ShiftSwapRequest"
+        )
+        failure_counter += 1
+
+    return new_public_primary_key
+
+
+class ShiftSwapRequestQueryset(models.QuerySet):
+    def delete(self):
+        self.update(deleted_at=timezone.now())
+
+
+class ShiftSwapRequestManager(models.Manager):
+    def get_queryset(self):
+        return ShiftSwapRequestQueryset(self.model, using=self._db).filter(deleted_at=None)
+
+    def hard_delete(self):
+        return self.get_queryset().hard_delete()
+
+
+class ShiftSwapRequest(models.Model):
+
+    objects = ShiftSwapRequestManager()
+    objects_with_deleted = models.Manager()
+
+    public_primary_key = models.CharField(
+        max_length=20,
+        validators=[MinLengthValidator(settings.PUBLIC_PRIMARY_KEY_MIN_LENGTH + 1)],
+        unique=True,
+        default=generate_public_primary_key_for_shift_swap_request,
+    )
+
+    created_at = models.DateTimeField(auto_now_add=True)
+    updated_at = models.DateTimeField(auto_now=True)
+    deleted_at = models.DateTimeField(null=True)
+
+    schedule = models.ForeignKey(
+        to="schedules.OnCallSchedule", null=False, on_delete=models.CASCADE, related_name="shift_swap_requests"
+    )
+
+    time_zone = models.CharField(max_length=100)
+
+    swap_start = models.DateTimeField()
+    swap_end = models.DateTimeField()
+
+    description = models.TextField(max_length=3000, default=None, null=True)
+
+    # beneficiary - the person who is relieved from (part of) their shift(s)
+    beneficiary = models.ForeignKey(
+        to="user_management.User", null=False, on_delete=models.CASCADE, related_name="created_shift_swap_requests"
+    )
+    # benefactor - the person taking on shift workload from the beneficiary
+    benefactor = models.ForeignKey(
+        to="user_management.User", null=True, on_delete=models.CASCADE, related_name="taken_shift_swap_requests"
+    )
+
+    OPEN = "open"
+    TAKEN = "taken"
+    PAST_DUE = "past_due"
+    DELETED = "deleted"
+
+    def __str__(self) -> str:
+        return (
+            f"{str(self.schedule.name)} {str(self.beneficiary.username)} {str(self.swap_start)} - {str(self.swap_end)}"
```

Comment:
calling `str` seems to be excessive here, `f` strings should handle that internally